### PR TITLE
[Core] Multiprocessing Pipeline Parallel support

### PR DESCRIPTION
**Pull request by njhill**
Only for single-node.

### CODE REVIEW

File: vllm/executor/multiproc_gpu_executor.py [hunk 1]
```diff
@@ -23,7 +24,8 @@ class MultiprocessingGPUExecutor(DistributedGPUExecutor):
 
     def _init_executor(self) -> None:
         # Create the parallel GPU workers.
-        world_size = self.parallel_config.tensor_parallel_size
+        world_size = self.parallel_config.pipeline_parallel_size \
```
Comment (andoorve) [inline vllm/executor/multiproc_gpu_executor.py:27]: We can just use `self.parallel_config.world_size`

File: vllm/executor/ray_gpu_executor.py [hunk 2]
```diff
@@ -344,11 +340,16 @@ class RayGPUExecutorAsync(RayGPUExecutor, DistributedGPUExecutorAsync):
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
         self.driver_exec_method = make_async(self.driver_worker.execute_method)
+        self.pp_locks: Optional[List[asyncio.Lock]] = None
 
     async def _driver_execute_model_async(
         self,
         execute_model_req: Optional[ExecuteModelRequest] = None
     ) -> List[SamplerOutput]:
+        if not self.tp_driver_workers:
```
Comment (andoorve) [inline vllm/executor/ray_gpu_executor.py:349]: This is a nice touch - we can probably add it in the Ray version as well

File: vllm/executor/multiproc_gpu_executor.py (hunk 1 above)
Comment (andoorve) [reply vllm/executor/multiproc_gpu_executor.py:27]: ^ nit

**Review by andoorve — APPROVED** [review]

File: vllm/executor/gpu_executor.py [hunk 3]
```diff
@@ -11,6 +11,15 @@
 logger = init_logger(__name__)
 
 
+def create_worker(worker_module_name, worker_class_name, **kwargs):
```
Comment (njhill) [inline vllm/executor/gpu_executor.py:14]: The changes in this file are to ensure that the function/args passed to the new worker process don't contain a reference to the executor instance itself (`self`).

File: vllm/executor/ray_gpu_executor.py (hunk 2 above)
Comment (njhill) [reply vllm/executor/ray_gpu_executor.py:349]: @andoorve this is the Ray version :-)

Comment (andoorve) [reply vllm/executor/ray_gpu_executor.py:349]: Ah of course!

File: vllm/executor/gpu_executor.py (hunk 3 above)
Comment (andoorve) [reply vllm/executor/gpu_executor.py:14]: What happens if we don't have this? Is this more of a code cleanliness thing?

**Comment by njhill** [comment]
There are some issues with the (daemon) miultiprocessing processes not being terminated in the CI, I think due to the test server that spawns them running in a ray actor. This is causing OOM as these processes accumulate. Need to look into this further.

**Comment by andoorve** [comment]
Hey @njhill any luck here?

File: vllm/executor/gpu_executor.py (hunk 3 above)
Comment (njhill) [reply vllm/executor/gpu_executor.py:15]: @andoorve no, for some reason there are pickling errors when forking the workers without this. What's confusing is why it works in the non-PP case.

File: vllm/executor/ray_gpu_executor.py [hunk 4]
```diff
@@ -240,27 +241,14 @@ def _init_workers_ray(self, placement_group: "PlacementGroup",
...
-        # Enforce rank order for correct rank to return final output.
-        self.tp_driver_workers = [
-            worker for _, worker in sorted(
-                zip(tp_driver_worker_ranks, self.tp_driver_workers))
-        ]
-        self.non_driver_workers = [
-            worker for _, worker in sorted(
-                zip(non_driver_worker_ranks, self.non_driver_workers))
-        ]
+                self.non_driver_workers.append(worker)
```
Comment (njhill) [inline vllm/executor/ray_gpu_executor.py:251]: Just a simplification to the logic here

**Comment by njhill** [comment]
@andoorve I think this is finally ready if you'd like to take another quick look, sorry for the delay!

File: vllm/executor/ray_gpu_executor.py (hunk 4 above)
Comment (andoorve) [reply vllm/executor/ray_gpu_executor.py:251]: Thanks! This makes a lot more sense aha

**Review by andoorve — APPROVED** [review]

**Review by andoorve — APPROVED** [review]
Thanks for your extensive refactoring! Looks great from my POV. The only request I have is if you could kick off the pipeline parallel tests 3-4 times if possible. We have been prone to some bugs that trigger intermittently, and want to make sure none of the refactoring affected it.

Pinging @youkaichao for approval as well

**Comment by njhill** [comment]
Thanks @andoorve! I'm actually not sure whether it's possible to re-trigger individual tests once they have passed :thinking:

File: vllm/executor/gpu_executor.py (hunk 3 above)
Comment (youkaichao) [reply vllm/executor/gpu_executor.py:15]: how does this solve the problem?

**Comment by andoorve** [comment]
I think this command works locally:

```
./.buildkite/run-multi-node-test.sh /vllm-workspace/tests 2 2 public.ecr.aws/q9t5s3a7/vllm-ci-test-repo:f0d5e5cdb0f9da33cfc7ac596de0be4155b594e9 "VLLM_TEST_SAME_HOST=0 torchrun --nnodes 2 --nproc-per-node=2 --rdzv_backend=c10d --rdzv_endpoint=192.168.10.10 distributed/test_same_node.py && pytest -v -s distributed/test_pipeline_parallel.py" "VLLM_TEST_SAME_HOST=0 torchrun --nnodes 2 --nproc-per-node=2 --rdzv_backend=c10d --rdzv_endpoint=192.168.10.10 distributed/test_same_node.py"
```

File: vllm/executor/multiproc_gpu_executor.py [hunk 5]
```diff
@@ -172,16 +198,45 @@ class MultiprocessingGPUExecutorAsync(MultiprocessingGPUExecutor,
...
+            # We create the locks here to avoid creating them in the constructor
+            # which uses a different asyncio loop.
+            self.pp_locks = [
+                asyncio.Lock()
+                for _ in range(self.parallel_config.pipeline_parallel_size)
+            ]
+
+        tasks = [
+            asyncio.create_task(
+                _run_task_with_lock(self.driver_exec_model, self.pp_locks[0],
```
Comment (youkaichao) [inline vllm/executor/multiproc_gpu_executor.py:222]: it always uses `self.pp_locks[0]`?

File: vllm/executor/gpu_executor.py (hunk 3 above)
Comment (youkaichao) [reply vllm/executor/gpu_executor.py:15]: okay, I think the problem is that we use `partial` later:

```python
partial(
                        self._create_worker,
                        rank=rank,
                        local_rank=rank,
                        distributed_init_method=distributed_init_method,
                    ))
```

**Review by youkaichao — APPROVED** [review]
LGTM, thanks for the PR!

As @andoorve suggests, please run the pp related tests locally for several times to make sure it is not flaky.

File: vllm/executor/gpu_executor.py (hunk 3 above)
Comment (njhill) [reply vllm/executor/gpu_executor.py:15]: @youkaichao yes and in particular the implicit reference to `self` held by the instance method it contains.

**Pull request merged**